lib/spack/spack/stage.py

```diff
@@ -725,6 +739,154 @@ def _add_to_root_stage(self):
+        if not current_ext_stages:
+            # no currently extant external stages, start enumerating with 'a'
+            return "a"
+        last = sorted(current_ext_stages, key=sort_key)[-1]
+        return inc(last.name)
+
+    def _teardown_remote_build_dir(self):
+        """Destroy external build tree if not keep-stage
+        Otherwise this is kept as usual"""
+
+        def teardown(pth: Path):
+            for sub_item in pth.iterdir():
+                if sub_item.is_dir():
+                    teardown(sub_item)
+                else:
+                    sub_item.unlink()
+            pth.rmdir()
+
+        if self._remote_build_dir.exists():
+            teardown(self._remote_build_dir)
+        self._remote_build_dir = None
+
+    def _destroy_remote(self):
+        # copy back to stage may fail in event of error, make sure we clean up the
+        # associated external build dir in that event unless we're keeping the
+        # parent stage on cleanup
+        # If remote stage is not set, we never created one, package is already
+        # installed and we should do nothing here
+        if self._remote_build_dir:
+            try:
+                self._remove_context_link()
+            finally:
+                self._teardown_remote_build_dir()
+
+    def contains(self, file_path):
+        """Returns true if file_path is in root stage
+        or the remote build dir"""
+        # Check if file is in root stage
+        cont = super(CMakeBuildStage, self).contains(file_path)
+        # if not, check if it's in remote build dir
+        if not cont:
+            cont = str(self._remote_build_dir) in file_path
+        return cont
+
+    def create(self):
+        if not self.created:
+            # Establish typical root stage
+            super(CMakeBuildStage, self).create()
+            # Now establish remote build stage
+            # or pick up extant stage from existing  symlink
+            if not self._root_stage_context.exists():
+                try:
+                    self._remote_build_dir = self._setup_remote_build_stage()
+                    self._establish_context_link()
+                except Exception:
+                    self._teardown_remote_build_dir()
+                    self.created = False
+                    raise
+            else:
+                self._remote_build_dir = self._root_stage_context.resolve()
+
+    def destroy(self):
+        self._destroy_remote()
+        super(CMakeBuildStage, self).destroy()
+
+    def path_rel_to_stage(self, glob_expr):
+        """Returns path of archive files relative to the root stage dir
+        from the remote build dir transparently as if there was a normal
+        build directory"""
+        rel_to_stage = os.path.relpath(
+            os.path.join(
+                self._root_stage_context, os.path.relpath(glob_expr, self._remote_build_dir)
+            ),
+            self.path,
+        )
```

Comment:
I'll add the replace as a comment. I'd be fine with the string comp if packages didn't define their own archive files, but since they can, there's little way to protect against differently formed strings until we convert to pathlib.